dotcom-rendering/src/model/enhanceTableOfContents.ts

```diff
@@ -0,0 +1,122 @@
+import { JSDOM } from 'jsdom';
+
+const isHeading = (element: CAPIElement, level: 'H2' | 'H3' | 'H4') => {
+	if (
+		element._type !==
+			'model.dotcomrendering.pageElements.TextBlockElement' &&
+		element._type !==
+			'model.dotcomrendering.pageElements.SubheadingBlockElement'
+	)
+		return false;
+	const frag = JSDOM.fragment(element.html);
+	return frag.firstElementChild?.nodeName === level;
+};
+
+const isH2 = (
+	element: CAPIElement,
+): element is SubheadingBlockElement | TextBlockElement => {
+	return isHeading(element, 'H2');
+};
+
+const isH3 = (
+	element: CAPIElement,
+): element is SubheadingBlockElement | TextBlockElement => {
+	return isHeading(element, 'H3');
+};
+
+const isH4 = (
+	element: CAPIElement,
+): element is SubheadingBlockElement | TextBlockElement => {
+	return isHeading(element, 'H4');
+};
+
+const extractText = (
+	element: SubheadingBlockElement | TextBlockElement,
+): string => {
+	const frag = JSDOM.fragment(element.html);
+	if (!frag.firstElementChild) return '';
+	return frag.textContent?.trim() ?? '';
+};
+
+const hasH2s = (blocks: Block[]) => {
+	return blocks.find((block) => block.elements.find(isH2));
+};
+
+const hasH3s = (blocks: Block[]) => {
+	return blocks.find((block) => block.elements.find(isH3));
+};
+
+const hasInteractiveContentsElement = (blocks: Block[]): boolean => {
+	return !!blocks.find((block) =>
+		block.elements.find(
+			(element) =>
+				element._type ===
+				'model.dotcomrendering.pageElements.InteractiveContentsBlockElement',
+		),
+	);
+};
+
+export const enhanceTableOfContents = (
+	format: CAPIFormat,
+	blocks: Block[],
+): TOCType[] | undefined => {
+	if (
+		format.design !== 'ArticleDesign' ||
+		hasInteractiveContentsElement(blocks)
+	) {
+		return undefined;
+	}
+
+	const tocs: TOCType[] = [];
+	if (hasH2s(blocks)) {
+		blocks.forEach((block) => {
+			block.elements.forEach((element) => {
+				if (isH2(element)) {
+					tocs.push({
+						id: element.elementId,
+						title: extractText(element),
+						nested: [],
+					});
+				}
+				if (isH3(element)) {
```

Comment:
I just caught up with Harry around this and he proposed simplifying things

If there are h2s and h3s then these are nested. Otherwise there is no nesting. I'll make those changes now